### PR TITLE
[inetstack] Enhancement: move to shared object

### PR DESCRIPTION
**Pull request by iyzhang**
This PR moves the inetstack to our new shared object abstraction. In particular, it introduces a shared object for each of the following data structures:
* ArpPeer
* Icmpv4Peer
* PassiveSocket
* TcpPeer
* TcpQueue
* ControlBlock
* UdpPeer
* SharedQueue
* UdpPeer
* UdpQueue
* TestRuntime
* Engine
* DummyRuntime

Each of these data structures contains state that is shared across coroutines. When possible, I have named the coroutines `*_coroutine` but generally I have not made changes to the Futures, as that will come in a new change. I've also not made any effort to organize the cloning of SharedObjects in anyway. That will also be done in a future PR once we have moved to a more object-oriented architecture

### CODE REVIEW

File: src/rust/catnip/runtime/mod.rs [hunk 1]
```diff
@@ -96,11 +96,11 @@ macro_rules! expect_zero {
...
-    pub link_addr: MacAddress,
-    pub ipv4_addr: Ipv4Addr,
-    pub arp_options: ArpConfig,
-    pub tcp_options: TcpConfig,
-    pub udp_options: UdpConfig,
+    link_addr: MacAddress,
+    ipv4_addr: Ipv4Addr,
+    arp_config: ArpConfig,
+    tcp_config: TcpConfig,
+    udp_config: UdpConfig,
```
Comment (ppenna) [inline src/rust/catnip/runtime/mod.rs:103]: Thanks.

File: src/rust/inetstack/protocols/udp/peer.rs [hunk 2]
```diff
@@ -167,213 +144,226 @@ impl<const N: usize> UdpPeer<N> {
...
+            return Err(Fail::new(libc::EADDRINUSE, &cause));
         }
+
         // Check if this is an ephemeral port or a wildcard one.
         if EphemeralPorts::is_private(addr.port()) {
             // Allocate ephemeral port from the pool, to leave  ephemeral port allocator in a consistent state.
             self.ephemeral_ports.alloc_port(addr.port())?
         } else if addr.port() == 0 {
-            // Allocate ephemeral port.
+            // Allocate ephemeral potransport.
```
Comment (ppenna) [inline src/rust/inetstack/protocols/udp/peer.rs:243]: Typo

File: src/rust/inetstack/protocols/udp/peer.rs [hunk 3]
```diff
@@ -88,18 +75,16 @@ const SEND_QUEUE_MAX_SIZE: usize = 1024;
 
 /// UDP Peer
 pub struct UdpPeer<const N: usize> {
-    /// Underlying runtime.
-    rt: Rc<dyn NetworkRuntime<N>>,
+    /// Shared Demikernel runtime.
+    runtime: SharedDemiRuntime,
+    /// Underlying transpotransport.
```
Comment (ppenna) [inline src/rust/inetstack/protocols/udp/peer.rs:80]: Typo

**Review by ppenna — APPROVED** [review]
This change was brutal.

Thanks for all this work :)

I did not spot any new bugs, but I'm heavily relying on our test suite to approve this.

File: src/rust/inetstack/protocols/udp/peer.rs (hunk 2 above)
Comment (iyzhang) [reply src/rust/inetstack/protocols/udp/peer.rs:243]: Thanks :)

**Pull request merged**